permchain/langgraph/__init__.py

```diff
@@ -0,0 +1,235 @@
+
+    def __init__(self, parent_key: str, condition: str):
+        self.parent_key = parent_key
+        self.condition = condition
+        super().__init__(f"{self.parent_key}.{self.condition}")
+
+    def get_runnable(self) -> Union[Runnable, Callable]:
+        # Only used for structure, so the runnable should never be called
+        raise NotImplementedError
+        
+
+class Conditional(LangGraphNode):
+
+    def __init__(self, key: str, conditional_edge_mapping: Dict[str, str], callable: Callable):
+        self.callable = callable
+        self.branches = []
+
+        for condition, output in conditional_edge_mapping.items():
+            self.branches.append(Branch(key, condition))
+            
+        self.conditional_edge_mapping = conditional_edge_mapping
+            
+        super().__init__(key)
+
+    def get_runnable(self) -> Union[Runnable, Callable]:
+        return self.callable
+    
+
+
+#########################################################
+#                    EDGE CLASSES                       #
+#########################################################
+
+class LangGraphEdge:
+    def __init__(self, start_key: str, end_key: str):
+        self.start_key = start_key
+        self.end_key = end_key
+    
+    def flow(self, node_map: Dict[str, LangGraphNode]):
+        return (
+            Channel.subscribe_to(self.start_key) | 
+            node_map[self.start_key].get_runnable() | 
+            Channel.write_to(self.end_key)
+        )
+
+
+class BranchEdge(LangGraphEdge):
+
+    def flow(self, node_map: Dict[str, LangGraphNode]):
+        # flow should skip over the branch edge
+        raise NotImplementedError
+
+        
+class ConditionalEdge(LangGraphEdge):
+    
+    def __init__(self, base_key: str, branch_key: str):
+        self.base_key = base_key
+        self.branch_key = branch_key
+        if not branch_key.startswith(base_key) or not branch_key[len(base_key)] == ".":
+            raise ValueError(f"Invalid branch edge from {base_key} to {branch_key}")
+            
+        super().__init__(base_key, branch_key)
+
+    def _branch(self, data, condition, mapping):
+        result = condition(data)
+        return Channel.write_to(mapping[result])
+
+    def flow(self, node_map: Dict[str, LangGraphNode]):
+        conditional_node = node_map[self.base_key]
+        
+        return (
+            Channel.subscribe_to(self.start_key) | 
+            (
+                lambda x: self._branch(
+                    x, 
+                    conditional_node.get_runnable(), 
+                    conditional_node.conditional_edge_mapping
+                )
+            )
+        )
+
+
+class Graph:
+
+    def __init__(self):
+        end_node = End()
+        self.nodes = {end_node.key: end_node}
+        self.edges = []
+        
+        # self.connections = {}
+        # self.branches = {}
+        self.entry_point: Optional[str] = None
+
+    def register_node(self, node: Actor):
```

Comment:
Could we rename `register` to `add`. It's semantically identical, and `add` is more friendly